File: implementations/flutter/design_patterns_flutter/lib/patterns/structural/proxy/proxy.dart
import 'dart:async';
import 'product.dart';
import 'subject.dart';
import 'real_subject.dart';

class CachingProductProxy implements ProductService {
  final HttpProductService _real;
  final Map<String, Product> _cache = {};
  final Map<String, Future<Product>> _inflight = {};
  int _hits = 0;
  int _misses = 0;

  CachingProductProxy([HttpProductService? real])
    : _real = real ?? HttpProductService();

  @override
  Future<Product> getProduct(String id) async {
    final cached = _cache[id];
    if (cached != null) {
      _hits++;
      return cached;
    }

    final pending = _inflight[id];
    if (pending != null) {
      _hits++; // treat shared inflight as hit
      return pending;
    }

    _misses++;
    final req = _real
        .getProduct(id)
        .then((p) {
          _cache[id] = p;
          _inflight.remove(id);
          return p;
        })
        .catchError((e) {
          _inflight.remove(id);
          throw e;
        });

    _inflight[id] = req;
    return req;
  }

  Map<String, dynamic> getStats() {
    return {
      'cacheHits': _hits,
      'cacheMisses': _misses,
      'networkCalls': _real.getNetworkHits(),
      'cacheSize': _cache.length,
      'inflight': _inflight.length,
    };
  }

  void clearCache() => _cache.clear();
}

File: implementations/react-native/design-patterns-rn/src/patterns/structural/proxy/proxy.ts
import { HttpProductService } from './real-subject';
import type { ProductService } from './subject';
import type { Product } from './type/type';

export class CachingProductProxy implements ProductService {
  private real: HttpProductService;
  private cache = new Map<string, Product>();
  private inflight = new Map<string, Promise<Product>>();
  private hits = 0;
  private misses = 0;

  constructor(real = new HttpProductService()) {
    this.real = real;
  }

  async getProduct(id: string): Promise<Product> {
    const cached = this.cache.get(id);
    if (cached) {
      this.hits++;
      return cached;
    }

    const pending = this.inflight.get(id);
    if (pending) {
      this.hits++; // treat shared inflight as hit
      return pending;
    }

    this.misses++;
    const req = this.real.getProduct(id).then(p => {
      this.cache.set(id, p);
      this.inflight.delete(id);
      return p;
    }).catch(err => {
      this.inflight.delete(id);
      throw err;
    });

    this.inflight.set(id, req);
    return req;
  }

  getStats() {
    return {
      cacheHits: this.hits,
      cacheMisses: this.misses,
      networkCalls: this.real.getNetworkHits(),
      cacheSize: this.cache.size,
      inflight: this.inflight.size,
    };
  }

  clearCache() { this.cache.clear(); }
}
